src/pal/src/thread/process.cpp

```diff
@@ -1624,10 +1667,95 @@ class PAL_RuntimeStartupHelper
         }
     }
 
-    PAL_ERROR InvokeStartupCallback()
+    PAL_ERROR IsCoreClrProcessReady(ProcessModules *module, bool *pCoreClrIsReady)
+    {
+        typedef DWORD PALAPI (*GetRVA_PFN)();
+        DWORD coreclrProcessIsReadyContents;
+        char dacModulePath[MAX_LONGPATH];
+        PAL_ERROR pe = NO_ERROR;
+        HANDLE hProcess = NULL;
+        void *dl_handle = NULL;
+        GetRVA_PFN getrva;
+        char *pszLast;
+        SIZE_T read;
+        LPCVOID address;
+
+        *pCoreClrIsReady = false;
+
+        if (!IsCoreClrModule(module->Name))
+        {
+            _ASSERT(pe == NO_ERROR);
+            goto exit;
+        }
+
+        pszLast = strrchr(module->Name, '/');
+        if (pszLast == NULL)
+        {
+            _ASSERT(!"IsCoreClrProcessReady: can find separator in coreclr path\n");
+            pe = ERROR_INVALID_PARAMETER;
+            goto exit;
+        }
+
+        strncpy_s(dacModulePath, MAX_LONGPATH, module->Name, pszLast - module->Name);
+        strcat_s(dacModulePath, MAX_LONGPATH, "/" MAKEDLLNAME_A("mscordaccore"));
+
+        dl_handle = dlopen(dacModulePath, RTLD_LAZY);
+        if (dl_handle == NULL)
+        {
+            WARN("dlopen() failed; dlerror says '%s'\n", dlerror());
+            pe = ERROR_MOD_NOT_FOUND;
+            goto exit;
+        }
+
+        // Load dac and call entry point to get g_coreclrProcessIsReady RVA
+        getrva = (GetRVA_PFN)dlsym(dl_handle, "GetCoreClrProcessIsReadyRVA");
+        if (getrva == NULL)
+        {
+            WARN("dlsym(GetCoreClrProcessIsReadyRVA) failed; dlerror says '%s'\n", dlerror());
+            pe = ERROR_MOD_NOT_FOUND;
+            goto exit;
+        }
+
+        // Calculate the address of the g_coreclrProcessIsReady
+        address = ((BYTE *)module->BaseAddress) + (*getrva)();
+
+        hProcess = OpenProcess(READ_CONTROL, FALSE, m_processId);
+        if (hProcess == NULL)
+        {
+            pe = GetLastError();
+            goto exit;
+        }
+
+        // Get the contents of g_coreclrProcessIsReady in the coreclr process. The errors from read memory
```

Comment:
Somewhere in the comments it would be good to add a general description enumerating all the different interesting windows of time and how this algorithm handles attaching during each of them. This comment here is handy for one aspect of it, but I didn't catch anything that gave a full description.